dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ZonedIntervalIterable.java

```diff
@@ -170,17 +164,24 @@ private ZonedIntervalIterable(
         "Start date %s must precede end date %s",
         queryLogStart,
         queryLogEnd);
-    this.start = queryLogStart.truncatedTo(unit);
-    this.end = queryLogEnd.truncatedTo(unit);
 
-    if (!start.equals(queryLogStart)) {
-      LOG.warn("Start time has been truncated to {}", start);
-    }
+    this.start = queryLogStart.truncatedTo(ChronoUnit.SECONDS);
+    this.end = truncate(start, queryLogEnd, duration);
+
     if (!end.equals(queryLogEnd)) {
       LOG.warn("End time has been truncated to {}", end);
     }
   }
 
+  private ZonedDateTime truncate(
+      @Nonnull ZonedDateTime start,
+      @Nonnull ZonedDateTime queryLogEnd,
+      @Nonnull Duration duration) {
+    long numberOfUnits =
+        Math.floorDiv(Duration.between(start, queryLogEnd).getSeconds(), duration.getSeconds());
+    return start.plus(Duration.ofSeconds(numberOfUnits * duration.getSeconds()));
```

Comment:
In the context of `dividesADayEvenly`, does it make sense to make truncation of the `queryLogEnd` timestamp dependent upon the `start` timestamp?